### PR TITLE
add ability to record and replay module for testing purpose

### PR DESCRIPTION
**Pull request by goneri**
Depends-On: https://github.com/ansible/ansible-zuul-jobs/pull/1619

Use lib ``placebo`` (https://pypi.org/project/placebo/) to record the API
interaction. This allow us to replay later a previous interaction offline
and cover some cases that are hard to reproduce in a CI.

### CODE REVIEW

File: tests/integration/targets/module_utils_botocore_recorder/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
```
Comment (tremble) [inline tests/integration/targets/module_utils_botocore_recorder/tasks/main.yml:1]: I don't think this file's being used?

File: changelogs/fragments/placebo_record.yaml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+- Ability to record and replay the API interaction of a module for testing purpose. Show case the feature with an example.
```
Comment (tremble) [inline changelogs/fragments/placebo_record.yaml:3]:
```suggestion
- Ability to record and replay the API interaction of a module for testing purpose. Show case the feature with an example (https://github.com/ansible-collections/amazon.aws/pull/998).
```

File: tests/integration/targets/module_utils_botocore_recorder/main.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+      amazon.aws.aws_az_info:
+      register: result
+    - debug: var=result
+    - name: Get called information
+      amazon.aws.aws_caller_info:
+      register: result
+    - debug: var=result
+    - assert:
+        that:
+         - lookup('ansible.builtin.env', '_ANSIBLE_PLACEBO_RECORD') or (lookup('ansible.builtin.env', '_ANSIBLE_PLACEBO_REPLAY') and result.user_id == "AWZBREIZHEOMABRONIFVGFS6GH")
```
Comment (tremble) [inline tests/integration/targets/module_utils_botocore_recorder/main.yml:14]:
```suggestion
    - assert:
        that:
        - '"user_id" in result'
        when:
        - lookup('ansible.builtin.env', '_ANSIBLE_PLACEBO_RECORD')
    - assert:
        that:
        - '"user_id" in result'
        # Once recorded we fully anonymize the UserID in the recording
        -  result.user_id == "AWZBREIZHEOMABRONIFVGFS6GH"
        when:        
        - lookup('ansible.builtin.env', '_ANSIBLE_PLACEBO_REPLAY')
    - assert:
        that:
        - False
        when:
        - not lookup('ansible.builtin.env', '_ANSIBLE_PLACEBO_REPLAY')
        - not lookup('ansible.builtin.env', '_ANSIBLE_PLACEBO_RECORD')
```

Do you think this is easier to follow as an example or am I over-thinking this?

**Review by tremble — COMMENTED** [review]
Thanks for splitting this off, generally LGTM but a couple of minor comments

File: tests/integration/targets/module_utils_botocore_recorder/main.yml (hunk 3 above)
Comment (goneri) [reply tests/integration/targets/module_utils_botocore_recorder/main.yml:14]: I actually get this error `"Invalid options for assert: when"`. I will keep the original version.

**Review by tremble — APPROVED** [review]
I'd love to see a follow up which includes some documentation around how to re-record.

We should also keep an eye on the size of any recordings.  Depending on how frequently the recordings are updated, this *could* explode the size of the repo without a clean way to shrink it back down.

**Comment by tremble** [comment]
@goneri I've not backported this to stable-4 since I don't think we'll have any specific needs for it, but I wouldn't block a backport if someone requested one.

**Pull request merged**